src/internal_modules/roc_core/memory_limiter.cpp

```diff
@@ -0,0 +1,73 @@
+/*
+ * Copyright (c) 2023 Roc Streaming authors
+ *
+ * This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this
+ * file, You can obtain one at http://mozilla.org/MPL/2.0/.
+ */
+
+#include "roc_core/memory_limiter.h"
+
+#include "roc_core/cpu_instructions.h"
+#include "roc_core/log.h"
+
+namespace roc {
+namespace core {
+
+MemoryLimiter::MemoryLimiter(const char* name, size_t max_bytes)
+    : name_(name)
+    , max_bytes_(max_bytes)
+    , bytes_acquired_(0) {
+}
+
+MemoryLimiter::~MemoryLimiter() {
+    if (bytes_acquired_ > 0) {
+        roc_panic("memory limiter (%s): detected that memory has not been released, "
+                  "currently acquired %lu",
+                  name_, (unsigned long)bytes_acquired_);
+    }
+}
+
+bool MemoryLimiter::acquire(size_t num_bytes) {
+    if (num_bytes == 0) {
+        roc_panic("memory limiter (%s): tried to acquire zero bytes", name_);
+    }
+    size_t current;
+    do {
+        current = bytes_acquired_;
+        size_t next = current + num_bytes;
+        if (max_bytes_ > 0 && next > max_bytes_) {
+            break;
+        }
+        if (bytes_acquired_.compare_exchange(current, next)) {
+            return true;
+        }
+        cpu_relax();
+    } while (true);
+    roc_log(LogError,
+            "memory limiter (%s): could not acquire %lu bytes due to limit, currently "
+            "acquired %lu, limit is %lu",
```

Comment:
nitpick: usually in this project logs has the following style:

```
component: text message: var=value var=value...
```

for example:

```
memory limiter (%s): could not acquire bytes due to limit: requested=%lu acquired=%lu limit=%lu
```

It's not strictly necessary, but recommended.

Such logs are easier to grep:

* you can grep entire "text message" in source code
* you can grep "var=" in console output

Reply:
Totally get the purpose. I wasn't sure we had a convention.